### ISSUE TITLE
First flushCommands call reading from unordered nodes may not work

### ISSUE THREAD

**Issue by wangkekekexili** [lettuce-io/lettuce-core#1658]
Not sure if it's a bug or I misconfigured anything. Any suggestion or advice is greatly appreciated!

I'm using Lettuce to connect to Redis cluster, and set auto flush commands to false. If it reads from unordered list of nodes, the first call is very likely to fail. 

#### Input Code

```kotlin
fun main(args: Array<String>) = runBlocking<Unit> {
    val redisEndpoint = "redis://localhost:30001"
    val clientResources = DefaultClientResources.builder()
        .dnsResolver(DirContextDnsResolver()).build()
    val redisClient = RedisClusterClient.create(clientResources, RedisURI.create(redisEndpoint))
    redisClient.setOptions(
        ClusterClientOptions
            .builder()
            .autoReconnect(true)
            .socketOptions(
                SocketOptions
                    .builder()
                    .connectTimeout(Duration.ofMillis(500))
                    .keepAlive(true)
                    .build()
            )
            .timeoutOptions(
                TimeoutOptions
                    .builder()
                    .fixedTimeout(Duration.ofMillis(2000L))
                    .build()
            )
            .topologyRefreshOptions(
                ClusterTopologyRefreshOptions
                    .builder()
                    .enableAllAdaptiveRefreshTriggers()
                    .build()
            )
            .build()
    )
    val commands = redisClient.connect().apply {
        readFrom = ReadFrom.ANY
    }.async().apply {
        setAutoFlushCommands(false)
    }
    val f1 = commands.get("hello world")
    commands.flushCommands()
    try {
        println("f1 ${f1.await()}")
    } catch (ex: Exception) {
        ex.printStackTrace()
    }

    val f2 = commands.get("hello world")
    commands.flushCommands()
    println("f2 ${f2.await()}")

    val f3 = commands.get("hello world")
    commands.flushCommands()
    println("f3 ${f3.await()}")
}
```

My local redis cluster setup is 3 instances each with 2 replicas:

```
cluster slots
1) 1) (integer) 5461
   2) (integer) 10922
   3) 1) "127.0.0.1"
      2) (integer) 30002
      3) "ca21277591e362cfad55875b46d28e88a3a501c4"
   4) 1) "127.0.0.1"
      2) (integer) 30004
      3) "84b06e13074c1ec124489212f2a1ad999efb22f9"
   5) 1) "127.0.0.1"
      2) (integer) 30007
      3) "7a429c0b876582e65985d7b63b812de6e75947c0"
2) 1) (integer) 0
   2) (integer) 5460
   3) 1) "127.0.0.1"
      2) (integer) 30001
      3) "2f7e930dfc30ce79f10b2f5c9e5d9ae3d6a001c4"
   4) 1) "127.0.0.1"
      2) (integer) 30008
      3) "7e16eb68d3862c04a1d21ce1ca7666de8f9a39a5"
   5) 1) "127.0.0.1"
      2) (integer) 30009
      3) "2b86e9931f4ec577c78ccf321f71b5e61f3936c9"
3) 1) (integer) 10923
   2) (integer) 16383
   3) 1) "127.0.0.1"
      2) (integer) 30003
      3) "21e3bd99a6564c99bb592823d1efa029fef51530"
   4) 1) "127.0.0.1"
      2) (integer) 30006
      3) "12a29b11ee955debcd12c733776c588e91eaa382"
   5) 1) "127.0.0.1"
      2) (integer) 30005
      3) "4f8e184a88a26d11ffce9e9d43aed165a74c60cf"
```

What I'm seeing is that the first call fails at a very high rate:

```
io.lettuce.core.RedisCommandTimeoutException: Command timed out after 2 second(s)
	at io.lettuce.core.ExceptionFactory.createTimeoutException(ExceptionFactory.java:51)
	at io.lettuce.core.protocol.CommandExpiryWriter.lambda$potentiallyExpire$0(CommandExpiryWriter.java:167)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
f2 null
f3 null
```

And I observed that the second call and the third call never failed.

If I switched from ReadFrom.ANY to ReadFrom.REPLICA, the first call never failed.

I doubt the issue happens because ReadFrom.ANY is unordered. I switched to use a custom ReadFrom (pasted below) and the issue also happened:

```kotlin
object ReadFromReplicaUnordered : ReadFrom() {
    override fun select(nodes: Nodes?): MutableList<RedisNodeDescription> {
        val result = ArrayList<RedisNodeDescription>()
        nodes!!.forEach {
            if (it.role == RedisInstance.Role.SLAVE) {
                result.add(it)
            }
        }
        return result
    }

    override fun isOrderSensitive() = false
}
```

#### Environment

- Lettuce version: 5.2.2.RELEASE
- Redis version: 6.0.5

**Comment by mp911de** [comment]
There are two effects here that come together. Cluster connection acquisition is asynchronous. When you use the async API, then commands are dispatched eventually. On the other hand, flush commands is synchronous. Flushing commands while the connection was not yet established can lead to a state that leaves commands unflushed as the connection creation isn't synchronized with flushing.

TL;DR: Make sure that all connections are initialized before using flush commands.

**Comment by wangkekekexili** [comment]
@mp911de Thanks for your reply. I understand now that what I did can have a race condition.

Maybe a dumb question: can I ask how a connection can be verified to have been initialized?

I'm trying to do something like this

```kotlin
commands.statefulConnection.partitions.forEach {
    println(commands.statefulConnection.getConnection(it.nodeId).isOpen)
}

val a1 = commands.get("1")
val a2 = commands.get("3")
val a3 = commands.get("4")
commands.flushCommands()
try {
    println("a1 ${a1.await()}")
    println("a2 ${a2.await()}")
    println("a3 ${a3.await()}")
} catch (ex: Exception) {
    ex.printStackTrace()
}
```

But still seeing error:

```
true
true
true
true
true
true
true
true
true
a1 null
io.lettuce.core.RedisCommandTimeoutException: Command timed out after 2 second(s)
	at io.lettuce.core.ExceptionFactory.createTimeoutException(ExceptionFactory.java:51)
	at io.lettuce.core.protocol.CommandExpiryWriter.lambda$potentiallyExpire$0(CommandExpiryWriter.java:167)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.DefaultEventExecutor.run(DefaultEventExecutor.java:66)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
```
